sdk/database/helper/dbutil/dbutil_test.go

```diff
@@ -62,3 +63,45 @@ func TestStatementCompatibilityHelper(t *testing.T) {
 		t.Fatalf("mismatch: %#v, %#v", expectedStatements3, statements3)
 	}
 }
+
+func TestQueryHelper(t *testing.T) {
+	data := map[string]string{
+		// These are typical keys you find in a data map used with QueryHelper
+		"username":   "hello",
+		"name":       "hello",
+		"password":   "world",
+		"expiration": "24h",
+	}
+	for _, tc := range []struct {
+		tpl      string
+		expected string
+	}{
+		{"", ""},
+		{"somedb://{{username}}:{{password}}@something", "somedb://hello:world@something"},
+		// Unknown placeholders pass through as is
+		{"user={{name}} other={{unknown}}", "user=hello other={{unknown}}"},
+		// Various incorrect delimiters pass through as is
+		{"{{{{{{{{", "{{{{{{{{"},
+		{"{{username}} {{incomplete", "hello {{incomplete"},
+		{"VALID UNTIL '{{expiration}}'; {{", "VALID UNTIL '24h'; {{"},
+		// This case tests whether `{{!{{password}}` successfully looks past the earlier unmatched {{
+		{"}}backwards{{!{{password}}!", "}}backwards{{!world!"},
+	} {
+		assert.Equal(t, tc.expected, QueryHelper(tc.tpl, data),
+			"template processing produced unexpected result")
+	}
+}
+
+// TestQueryHelper_Recursion confirms QueryHelper does not replace placeholders that were themselves added as part of
+// a replacement value.
+func TestQueryHelper_Recursion(t *testing.T) {
+	data := map[string]string{
+		"a": "A{{a}}{{b}}{{c}}{{d}}",
+		"b": "B{{a}}{{b}}{{c}}{{d}}",
+		"c": "C{{a}}{{b}}{{c}}{{d}}",
+		"d": "D{{a}}{{b}}{{c}}{{d}}",
+	}
+	assert.Equal(t, "A{{a}}{{b}}{{c}}{{d}}B{{a}}{{b}}{{c}}{{d}}C{{a}}{{b}}{{c}}{{d}}D{{a}}{{b}}{{c}}{{d}}",
+		QueryHelper("{{a}}{{b}}{{c}}{{d}}", data),
+		"template processing produced unexpected result")
+}
```

Comment:
This test fails when applied to the existing implementation of `QueryHelper` - it showcases the issues just discussed at the end of #13414.